zinc/src/main/scala/sbt/internal/inc/MixedAnalyzingCompiler.scala

```diff
@@ -162,6 +165,17 @@ final class MixedAnalyzingCompiler(
  * of cross Java-Scala compilation.
  */
 object MixedAnalyzingCompiler {
+  // For more safety, store both the time and size
+  private type JarMetadata = (FileTime, Long)
+  private[this] val cacheMetadataJar = new ConcurrentHashMap[File, (JarMetadata, FileHash)]()
```

Comment:
Given that https://github.com/sbt/zinc/pull/427 is in, it would seem that all of this code should be located on the companion object of the "default" implementation of hashing. Put it in the top of `MixedAnalyzingCompiler` makes that class into more of a grabbag.

Reply:
By the companion object of the default implementation of hashing, do you mean `FileHash` or `Stamper`? It's not clear to me we want to reuse this code across all the users of `IOHash` or `FileHash`.

Reply:
If there isn't a relevant companion class, then I guess my suggestion would be to define a new one. Not a blocker. Just that logically I would expect this to look something like:
```
val hashingStrategy = if (flag) DefaultStrategy else UserDefinedStrategy(...)
???
```

But looking at it again, ExternalHooks doesn't make it possible to do this. So only the "grabbag" part is really relevant.

Reply:
I agree it looks like a grabbag, so it's better to move it somewhere else. thanks for the suggestion :smile:.

Reply:
@stuhood I've implemented your suggestion :wink:. The Zinc API has been historically such a mess that sometimes I forget it can and should be improved :smile:.